From 6427e49b4411b4368d79a6b5be06104fe699cdf4 Mon Sep 17 00:00:00 2001
From: Riddle Hsu <riddlehsu@google.com>
Date: Mon, 19 May 2025 15:07:16 +0800
Subject: [PATCH] Use consistent animation when forcibly hiding non system
 overlay

There are several non system overlay window types. When hiding the
windows with animation, their specified exit animations may be used.

This avoids ugly jump cut and inconsistent duration when these
windows are requested to be hidden.

Bug: 408215749
Flag: EXEMPT bugfix
Test: Show multiple windows with different overlay types and
      window animations by non system apps. Launch Settings
      and check the windows are fading out in a short time.
Merged-In: Ie7bb2663892d608715aa077e2170eae4c03a4e36
Change-Id: Ie7bb2663892d608715aa077e2170eae4c03a4e36
---
 services/core/java/com/android/server/wm/WindowState.java    | 4 ++++
 .../core/java/com/android/server/wm/WindowStateAnimator.java | 5 +++++
 2 files changed, 9 insertions(+)

diff --git a/services/core/java/com/android/server/wm/WindowState.java b/services/core/java/com/android/server/wm/WindowState.java
index 56b597e7dccc..758b813fa029 100644
--- a/services/core/java/com/android/server/wm/WindowState.java
+++ b/services/core/java/com/android/server/wm/WindowState.java
@@ -3334,6 +3334,10 @@ class WindowState extends WindowContainer<WindowState> implements WindowManagerP
         return true;
     }
 
+    boolean isForceHiddenNonSystemOverlayWindow() {
+        return mForceHideNonSystemOverlayWindow;
+    }
+
     void setForceHideNonSystemOverlayWindowIfNeeded(boolean forceHide) {
         final int baseType = getBaseType();
         if (mSession.mCanAddInternalSystemWindow
diff --git a/services/core/java/com/android/server/wm/WindowStateAnimator.java b/services/core/java/com/android/server/wm/WindowStateAnimator.java
index 5c0557f2d1f4..b6dd7cabe1a7 100644
--- a/services/core/java/com/android/server/wm/WindowStateAnimator.java
+++ b/services/core/java/com/android/server/wm/WindowStateAnimator.java
@@ -57,6 +57,7 @@ import android.view.Surface.OutOfResourcesException;
 import android.view.SurfaceControl;
 import android.view.WindowManager;
 import android.view.WindowManager.LayoutParams;
+import android.view.animation.AlphaAnimation;
 import android.view.animation.Animation;
 import android.view.animation.AnimationUtils;
 
@@ -623,6 +624,10 @@ class WindowStateAnimator {
                     a = AnimationUtils.loadAnimation(mContext, anim);
                     Trace.traceEnd(Trace.TRACE_TAG_WINDOW_MANAGER);
                 }
+            } else if (!isEntrance && mWin.isForceHiddenNonSystemOverlayWindow()) {
+                a = new AlphaAnimation(1f, 0f);
+                a.setDuration(mContext.getResources().getInteger(
+                        com.android.internal.R.integer.config_shortAnimTime));
             } else {
                 switch (transit) {
                     case WindowManagerPolicy.TRANSIT_ENTER:
-- 
2.50.1.351.gcf940e82a1-goog

